### ISSUE TITLE
sharding advice

### ISSUE THREAD

**Issue by d-v-b** [google/tensorstore#13]
I'm trying to save a large uint8 array with dimensions == (7500, 3400, 10391) as a multiresolution pyramid in the precomputed sharded format with the following spec:

```json
{
    "@type": "neuroglancer_multiscale_volume",
    "data_type": "uint8",
    "num_channels": 1,
    "scales": [
        {
            "chunk_sizes": [
                [
                    96,
                    96,
                    96
                ]
            ],
            "encoding": "jpeg",
            "jpeg_quality": 90,
            "key": "s0",
            "resolution": [
                4.0,
                4.0,
                4.28
            ],
            "sharding": {
                "@type": "neuroglancer_uint64_sharded_v1",
                "data_encoding": "raw",
                "hash": "identity",
                "minishard_bits": 6,
                "minishard_index_encoding": "gzip",
                "preshift_bits": 9,
                "shard_bits": 15
            },
            "size": [
                7500,
                3400,
                8158
            ],
            "voxel_offset": [
                0,
                0,
                0
            ]
        },
        {
            "chunk_sizes": [
                [
                    96,
                    96,
                    96
                ]
            ],
            "encoding": "jpeg",
            "jpeg_quality": 90,
            "key": "s1",
            "resolution": [
                8.0,
                8.0,
                8.56
            ],
            "sharding": {
                "@type": "neuroglancer_uint64_sharded_v1",
                "data_encoding": "raw",
                "hash": "identity",
                "minishard_bits": 6,
                "minishard_index_encoding": "gzip",
                "preshift_bits": 9,
                "shard_bits": 15
            },
            "size": [
                3750,
                1700,
                4079
            ],
            "voxel_offset": [
                0,
                0,
                0
            ]
        },
        {
            "chunk_sizes": [
                [
                    96,
                    96,
                    96
                ]
            ],
            "encoding": "jpeg",
            "jpeg_quality": 90,
            "key": "s2",
            "resolution": [
                16.0,
                16.0,
                17.119999999999997
            ],
            "sharding": {
                "@type": "neuroglancer_uint64_sharded_v1",
                "data_encoding": "raw",
                "hash": "identity",
                "minishard_bits": 6,
                "minishard_index_encoding": "gzip",
                "preshift_bits": 9,
                "shard_bits": 15
            },
            "size": [
                1875,
                850,
                2040
            ],
            "voxel_offset": [
                0,
                0,
                0
            ]
        },
        {
            "chunk_sizes": [
                [
                    96,
                    96,
                    96
                ]
            ],
            "encoding": "jpeg",
            "jpeg_quality": 90,
            "key": "s3",
            "resolution": [
                32.0,
                32.0,
                34.239999999999995
            ],
            "sharding": {
                "@type": "neuroglancer_uint64_sharded_v1",
                "data_encoding": "raw",
                "hash": "identity",
                "minishard_bits": 6,
                "minishard_index_encoding": "gzip",
                "preshift_bits": 9,
                "shard_bits": 15
            },
            "size": [
                938,
                425,
                1020
            ],
            "voxel_offset": [
                0,
                0,
                0
            ]
        },
        {
            "chunk_sizes": [
                [
                    96,
                    96,
                    96
                ]
            ],
            "encoding": "jpeg",
            "jpeg_quality": 90,
            "key": "s4",
            "resolution": [
                64.0,
                64.0,
                68.48000000000002
            ],
            "sharding": {
                "@type": "neuroglancer_uint64_sharded_v1",
                "data_encoding": "raw",
                "hash": "identity",
                "minishard_bits": 6,
                "minishard_index_encoding": "gzip",
                "preshift_bits": 9,
                "shard_bits": 15
            },
            "size": [
                469,
                213,
                510
            ],
            "voxel_offset": [
                0,
                0,
                0
            ]
        },
        {
            "chunk_sizes": [
                [
                    96,
                    96,
                    96
                ]
            ],
            "encoding": "jpeg",
            "jpeg_quality": 90,
            "key": "s5",
            "resolution": [
                128.0,
                128.0,
                136.96
            ],
            "sharding": {
                "@type": "neuroglancer_uint64_sharded_v1",
                "data_encoding": "raw",
                "hash": "identity",
                "minishard_bits": 6,
                "minishard_index_encoding": "gzip",
                "preshift_bits": 9,
                "shard_bits": 15
            },
            "size": [
                235,
                107,
                255
            ],
            "voxel_offset": [
                0,
                0,
                0
            ]
        }
    ],
    "type": "image"
}
```

Although the save completes without error, the resulting highest-resolution array seems incomplete when I view it with neuroglancer -- there are a lot of `Error retrieving chunk` errors emitted by the chunk worker, and a the total file size of the highest resolution array is clearly too small. 

I'm saving the data to a network store using nfs, which might be causing problems if worker state is being synchronized via the file system. I will try saving the same dataset on a different file system and see if the problems are the same.

But beyond trying a different filesystem, do you have any advice for debugging this?

**Comment by jbms** [comment]
If you are writing it in parallel from multiple machines over NFS, and your NFS server does not support byte range locking, then you could definitely run into problems.  If you carefully divide the work into blocks such that each block exactly aligns with a full shard, then you would also avoid the need for working locking, and it will also be much more efficient.  I think Stephen Plaza already has some code to do this.

If that doesn't prove to be the problem, if you can send me code to reproduce it on a local filesystem I can try to look into it.

**Comment by d-v-b** [comment]
How can I know in advance which chunks map to which shard files?

**Comment by jbms** [comment]
Unfortunately TensorStore doesn't yet provide an API at the moment for determining rectangular regions that correspond to full shards, or for automatically picking suitable sharding parameters, but the basic idea is as follows:

1. Use identity hash
2. Choose preshift_bits and minishard_bits such that `preshift_bits + minishard_bits == k * 3`.  This isn't strictly necessary but simplifies figuring out the size of the region corresponding to each shard.
3. Pick shard_bits such that `minishard_bits + preshift_bits + shard_bits` is exactly `sum_i ceil(log_2(shape[i] / chunk_shape[i]))`, where `shape` is the full volume shape and `chunk_shape` is the chunk shape.

Then every shard corresponds to a rectangular grid cell size `k * chunk_shape`, and you can compute the mapping from shard number to grid cell according to the documentation here: 
https://github.com/google/neuroglancer/blob/master/src/neuroglancer/datasource/precomputed/volume.md#compressed-morton-code

**Comment by d-v-b** [comment]
> Unfortunately TensorStore doesn't yet provide an API at the moment for determining rectangular regions that correspond to full shards, or for automatically picking suitable sharding parameters

This would be great to see. It would also be nice to have a method for converting a non-sharded store to a sharded store, which would allow mixing different parallelism strategies.

At the moment I'm resorting to computing small slabs of my image data in parallel and then saving them to a sharded store sequentially (on local storage). But this might be too slow to be practical (~1 day to save a (10,000,10,000,10,000)-sized dataset).

**Comment by chrisroat** [comment]
> Then every shard corresponds to a rectangular grid cell size `k * chunk_shape`

Do you mean `2**k * chunk_shape`?

I was just educating myself on the sharded format.  What is the purpose of preshift bits, exactly?

**Comment by jbms** [comment]
Sorry, yes, `2**k * chunk_shape`.  The purpose of `preshift_bits` is to allow you to have a certain number of nearby chunks (in compressed morton code space, which are also nearby in 3-d space) within the same minishard, due to the way the minishard and shard number is computed.

It has taken a while, but we are actually about to push out a large TensorStore change that has been in progress for a while, that makes it much easier to work with this format --- it will make it possible to query the 3-d rectangular shape of a shard, and to automatically choose sharding parameters that produce the desired shard shape.

**Comment by chrisroat** [comment]
I'm starting to understand, though I initially thought a minishard was a chunk.   It's starting to make sense now.  I would get one chunk per minishard with preshift_bits=0 (and identity hash).

**Comment by jbms** [comment]
The sharded format is a little weird --- I came up with it rather quickly, and wanted it to work for multiple purposes.

If you set shard_bits = 0, preshift_bits =0, and minishard_bits =0, then you actually end up with everything in the same minishard.  But normally we would choose shard_bits + minishard_bits + preshift_bits to be >= total number of bits in the chunk morton codes, so that higher-order bits aren't an issue.

**Comment by chrisroat** [comment]
OK, I'm starting to understand better, I think.  I'm using dask to parallelize over my volumes, and usually choose a dask block around 100MB, so something like 512*512*256 (uint16) would be the shard size.   I imagine most people write shards at a larger scale -- I could always merge them into larger blocks before writing.

With chunks of 64x64x64, I would have 256 chunks per shard.   Would a logarithmically balanced hierarchy of ​N chunks/minishard and N minishards/shard make sense (N=16 for my example)?

**Comment by jbms** [comment]
With only 256 chunks per shard, you would be fine to have just a single minishard, I think.

**Comment by chrisroat** [comment]
What are typical shard sizes, # minishards, and # chunks that you use in volumetric data?  I may merge chunks for I/O if it makes more sense.

**Comment by jbms** [comment]
I typically aim for 256 to 512 chunks per minishard for volumetric data.  As for how the minishards are grouped into shards, that doesn't affect read performance at all.

More shards -> greater parallelism when writing.

Fewer shards -> easier to copy/move the data around.

I think if you aim for at least 100MB shards you should be fine.
